// @threlte/core/dist/context/fragments/renderer.svelte.js
import { getContext, setContext, untrack } from 'svelte';
import { AgXToneMapping, ColorManagement, SRGBColorSpace, PCFSoftShadowMap, WebGLRenderer, OrthographicCamera, PerspectiveCamera } from 'three';
import { runeToCurrentReadable, runeToCurrentWritable } from '../../utilities/currentWritable.js';
import { useCamera } from './camera.svelte.js';
import { useDOM } from './dom.svelte.js';
import { useScene } from './scene.js';
import { useScheduler } from './scheduler.svelte.js';
import { fromStore } from 'svelte/store';
import { devicePixelRatio } from 'svelte/reactivity/window';
import { updateCamera } from '../../components/T/utils/useCamera.svelte.js';
export const createRendererContext = (options) => {
    const { camera, manual } = useCamera();
    const { scene } = useScene();
    const { invalidate, mainStage, renderStage, autoRender: autoRenderStore, scheduler, frameInvalidated } = useScheduler();
    const { canvas, size, shouldUpdateSize } = useDOM();
    const opts = $derived(options());
    const renderer = untrack(() => opts.createRenderer
        ? opts.createRenderer(canvas)
        : new WebGLRenderer({
            canvas,
            powerPreference: 'high-performance',
            antialias: true,
            alpha: true
        }));
    const resizeStage = scheduler.createStage(Symbol('threlte-resize-stage'), {
        before: mainStage
    });
    resizeStage.createTask(Symbol('threlte-resize-task'), () => {
        if (renderer.xr.isPresenting)
            return;
        if (!shouldUpdateSize())
            return;
        renderer.setSize(size.current.width, size.current.height);
        if (!manual.current) {
            updateCamera(camera.current, size.current.width, size.current.height);
        }
        invalidate();
    });
    const autoRenderTask = renderStage.createTask(Symbol('threlte-auto-render-task'), () => {
        renderer.render(scene, camera.current);
    });
    const optsColorSpace = $derived(opts.colorSpace);
    const optsDpr = $derived(opts.dpr);
    const optsShadows = $derived(opts.shadows);
    const optsToneMapping = $derived(opts.toneMapping);
    // Seperate derived runes since users can set these values through the canvas or by .set()
    let colorSpace = $derived(optsColorSpace ?? SRGBColorSpace);
    let dpr = $derived.by(() => {
        const target = devicePixelRatio.current ?? window.devicePixelRatio;
        if (Array.isArray(optsDpr)) {
            return Math.min(Math.max(optsDpr[0], target), optsDpr[1]);
        }
        return optsDpr ?? target;
    });
    let shadows = $derived(optsShadows ?? PCFSoftShadowMap);
    let toneMapping = $derived(optsToneMapping ?? AgXToneMapping);
    const context = {
        renderer: renderer,
        colorManagementEnabled: runeToCurrentReadable(() => opts.colorManagementEnabled ?? true),
        colorSpace: runeToCurrentWritable(() => colorSpace, (value) => (colorSpace = value)),
        dpr: runeToCurrentWritable(() => dpr, (value) => (dpr = value)),
        shadows: runeToCurrentWritable(() => shadows, (value) => (shadows = value)),
        toneMapping: runeToCurrentWritable(() => toneMapping, (value) => (toneMapping = value)),
        autoRenderTask
    };
    setContext('threlte-renderer-context', context);
    const autoRender = fromStore(autoRenderStore);
    ColorManagement.enabled = opts.colorManagementEnabled ?? true;
    $effect.pre(() => {
        renderer.outputColorSpace = colorSpace;
        invalidate();
    });
    $effect.pre(() => {
        renderer.setPixelRatio(dpr);
        invalidate();
    });
    $effect.pre(() => {
        renderer.shadowMap.enabled = shadows !== false;
        // If shadows isn't a boolean, it's a shadowMap type
        if (shadows !== false && shadows !== true) {
            renderer.shadowMap.type = shadows;
        }
        else if (shadows === true) {
            renderer.shadowMap.type = PCFSoftShadowMap;
        }
        invalidate();
    });
    $effect.pre(() => {
        renderer.toneMapping = toneMapping;
        invalidate();
    });
    $effect.pre(() => {
        if (autoRender.current) {
            context.autoRenderTask.start();
        }
        else {
            context.autoRenderTask.stop();
        }
        invalidate();
        return () => {
            context.autoRenderTask.stop();
        };
    });
    renderer.setAnimationLoop((time) => {
        scheduler.run(time);
        frameInvalidated.current = false;
    });
    $effect(() => {
        return () => {
            renderer.setAnimationLoop(null);
            try {
                renderer.dispose();
            }
            catch {
                // WebGPURenderer.dispose() throws if async init() hasn't completed (e.g. during HMR)
            }
        };
    });
    return context;
};
export const useRenderer = () => {
    const context = getContext('threlte-renderer-context');
    if (!context) {
        throw new Error('useRenderer can only be used in a child component to <Canvas>.');
    }
    return context;
};
